### PR TITLE
[SPARK-23488][python] Add missing catalog methods to python API

### PR DESCRIPTION
**Pull request by drboyer**
## What changes were proposed in this pull request?

As noted in SPARK-23488, the Python Catalog API was missing some methods that are present in the Scala API. Both for the sake of consistency, and because of their utility, I've added the missing methods (the database/table/functionExists() and getDatabase/Table/Function() methods).

I modeled these methods off of how the existing ones were written in catalog.py.

## How was this patch tested?

manually tested the added methods and compared functionality to the scala API counterparts

## Questions

I wasn't sure whether to set the `@since(x.y)` decorators on the new functions, and if so, what value to set them to?

### CODE REVIEW

**Review by HyukjinKwon — COMMENTED** [review]
Can we have a doctest and unittests too?

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by drboyer** [comment]
@HyukjinKwon thanks for the review so far! Sorry for the delay, I somehow missed the Python style output in the test logs earlier. How's this look now?

Can you elaborate more on "doctest" if it's still needed? From what I can tell the only documentation for the Catalog is [this simple reference](https://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=catalog#pyspark.sql.SparkSession.catalog) which would be unaffected by my change

**Comment by HyukjinKwon** [comment]
Actually, I think I need some more feedback about matching https://github.com/apache/spark/pull/20658#discussion_r170003442. 

@rxin, @hvanhovell and @cloud-fan (from checking the history), 

the problem here seems Scala side of `Function`:

https://github.com/apache/spark/blob/04d417a7ca8ef694658b26fb697a035717414731/sql/core/src/main/scala/org/apache/spark/sql/catalog/interface.scala#L129-L135

has `database`. Was checking the history. Looks this was added later alone and was missed in Python one:

https://github.com/apache/spark/blob/a49ffa010a46a0d87de124d8ddf66c8173b756fb/python/pyspark/sql/catalog.py#L31

We can't directly add it because it breaks compatibility, for example, if we go as below:

```
Function = namedtuple("Function", "name database description className isTemporary") 
```

It breaks

```python
>>> Function("a","b", "c", "d")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __new__() takes exactly 6 arguments (5 given)
```

We could workaround it by a hack to inject a default value to allow the case above but it is still a problem, for example, if we call like `len` and then length will be changed.

Do you guys think it's okay to just add it as below with a release/migration note?

```diff
- Function = namedtuple("Function", "name description className isTemporary") 
+ Function = namedtuple("Function", "name database description className isTemporary") 
```

**Comment by gatorsmile** [comment]
cc @ueshin

**Comment by holdenk** [comment]
So it looks like adding the database field has been backed out for this change, which is probably the right thing to do, if we want to make that change we should do that in a separate JIRA & commit so it can be rolled back separately from the new methods.

**Review by holdenk — COMMENTED** [review]
This looks like a good start. To make it a bit better we could use the \:param\: syntax in the docstring (see registerFunction for an example in catalog.py). This is a minor point though.

**Comment by drboyer** [comment]
Thanks @holdenk, I can open a separate JIRA about the missing field in `Function` if it seems worth fixing. It wasn't critical for me, I just happened to notice while doing some testing so I included it in my inital commit.

I hadn't added more complex docstrings just since these seemed like pretty simple methods with straightforward parameters. Happy to add :param: and :return: annotations if desired, but should we add these to some of the other catalog methods as well if we're adding it to these new ones (thinking especially of the `list*()` methods)?

File: python/pyspark/sql/catalog.py [hunk 1]
```diff
@@ -137,6 +138,78 @@ def listColumns(self, tableName, dbName=None):
                 isBucket=jcolumn.isBucket()))
         return columns
 
+    @ignore_unicode_prefix
+    # TODO: @since() decorator?
```
[the remark this answers is not in the document]
Comment (holdenk) [reply python/pyspark/sql/catalog.py]: Now 3.0

**Comment by holdenk** [comment]
If you have time to update this to master (either merge or rebase is fine), that would be great. In the meantime Jenkins ok to test.

**Comment by gatorsmile** [comment]
ping @drboyer 

Could either @deepyaman , @icexelloss or @huaxingao help fix this?

**Comment by HyukjinKwon** [comment]
This needs a bit of breaking change but that's quite unlikely corner case. Since we're going ahead for Spark 3.0, I think we can fix it. Please address https://github.com/apache/spark/pull/20658#issuecomment-368561007 and take this over if the author is being active.

**Comment by drboyer** [comment]
Hi, sorry for the delay and thanks all for taking a second look at this. I just rebased to latest master and pushed some related changes, like setting the `@since` decorator to `3.0`.

I see I never opened the Jira about the breaking `Function` change mentioned in @HyukjinKwon's [comment](https://github.com/apache/spark/pull/20658#issuecomment-368561007). I've opened [SPARK-26505](https://issues.apache.org/jira/browse/SPARK-26505) about this. Should that change be made in a separate PR or should we re-add that change here, now that it's targeted for inclusion in 3.0?

From my perspective, the only other thing discussed in previous reviews was potentially improving the docstrings for catalog functions, including the ones added here, to include :param:/:return: annotations.  Should I go ahead and make those changes too? Is there anything else needed before this can be merged?

**Comment by HyukjinKwon** [comment]
And, yea, let's add docstrings. I don't know why this APIs doesn't have any doctest or argument descriptions. It should have.

**Comment by HyukjinKwon** [comment]
Yea yea. I also think other docs should be done in a separate PR. Sorry that I wasn't clear.

File: python/pyspark/sql/catalog.py [hunk 2]
```diff
@@ -138,6 +139,78 @@ def listColumns(self, tableName, dbName=None):
...
 
+    @ignore_unicode_prefix
+    @since(3.0)
+    def databaseExists(self, dbName):
+        """Check if the database with the specified name exists."""
+        return self._jcatalog.databaseExists(dbName)
+
+    @ignore_unicode_prefix
+    @since(3.0)
+    def functionExists(self, functionName, dbName=None):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: Hm, this signature looks different with Scala side:

```scala
  def functionExists(functionName: String): Boolean
  def functionExists(dbName: String, functionName: String): Boolean
```

Can we handle this? There're some examples to refer, for instance,

https://github.com/apache/spark/blob/ecaa495b1fe532c36e952ccac42f4715809476af/python/pyspark/sql/dataframe.py#L811-L846

File: python/pyspark/sql/catalog.py [hunk 3]
```diff
@@ -138,6 +139,78 @@ def listColumns(self, tableName, dbName=None):
...
+
+        If no database is specified, the current database is used.
+        """
+        if dbName is None:
+            dbName = self.currentDatabase()
+        return self._jcatalog.functionExists(dbName, functionName)
+
+    @ignore_unicode_prefix
+    @since(3.0)
+    def tableExists(self, tableName, dbName=None):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: ditto for matching signautre.

File: python/pyspark/sql/catalog.py [hunk 4]
```diff
@@ -138,6 +139,78 @@ def listColumns(self, tableName, dbName=None):
...
+        """Get the database with the specified name."""
+        database = self._jcatalog.getDatabase(dbName)
+        return Database(
+            name=database.name(),
+            description=database.description(),
+            locationUri=database.locationUri())
+
+    @ignore_unicode_prefix
+    @since(3.0)
+    def getTable(self, tableName, dbName=None):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: ditto.

File: python/pyspark/sql/catalog.py [hunk 5]
```diff
@@ -138,6 +139,78 @@ def listColumns(self, tableName, dbName=None):
...
+        return Table(
+            name=table.name(),
+            database=table.database(),
+            description=table.description(),
+            tableType=table.tableType(),
+            isTemporary=table.isTemporary())
+
+    @ignore_unicode_prefix
+    @since(3.0)
+    def getFunction(self, functionName, dbName=None):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: ditto

File: python/pyspark/sql/catalog.py [hunk 6]
```diff
@@ -138,6 +139,78 @@ def listColumns(self, tableName, dbName=None):
...
+        """Get the function with the specified name.
+
+        If no database is specified, the current database is used.
+        """
+        if dbName is None:
+            dbName = self.currentDatabase()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: Can you check this logic? In Scala side, it seeks the table and set the database if possible whereas PySpark version only set it as currentDatabase. We can do for instance:

```python
args = []
if dbName is set:
  args += dbName
if function name is set:
  args += functionName
self._jcatalog.getFunction(*args)
```

File: python/pyspark/sql/catalog.py [hunk 7]
```diff
@@ -138,6 +139,78 @@ def listColumns(self, tableName, dbName=None):
...
+    def databaseExists(self, dbName):
+        """Check if the database with the specified name exists."""
+        return self._jcatalog.databaseExists(dbName)
+
+    @ignore_unicode_prefix
+    @since(3.0)
+    def functionExists(self, functionName, dbName=None):
+        """Check if the function with the specified name exists.
+
+        If no database is specified, the current database is used.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: Can we add doctests?

**Comment by HyukjinKwon** [comment]
ping @drboyer

File: python/pyspark/sql/catalog.py (hunk 2 above)
Comment (holdenk) [reply python/pyspark/sql/catalog.py]: I think as written is more Python-standard API. Certainly we can do interesting things with vargs/kwargs to make this look more like Scala, but I don't think that would be ideal here. One possibility could be requiring only kwargs (named arguments) so it's not ambigious. What do you think?

File: python/pyspark/sql/catalog.py (hunk 3 above)
Comment (holdenk) [reply python/pyspark/sql/catalog.py]: Same comment, I don't think matching the Scala API is necessarily the right thing to do here. I think this more closely matches Python developers expectations of positional arguments behaviour.

File: python/pyspark/sql/catalog.py (hunk 4 above)
Comment (holdenk) [reply python/pyspark/sql/catalog.py]: same

**Review by holdenk — COMMENTED** [review]
I have a different opinion about handling the positional arguments here to match what is more common in Python. In December I think we agreeed that provided it's tested in a regular test file it was OK to defer the doctests to a follow up PR, do we want to do the doctests now instead?

**Comment by holdenk** [comment]
Jenkins ok to test

**Comment by HyukjinKwon** [comment]
Since we're going ahead for 3.0, I think I am fine with the positional argument stuff now ..

**Comment by HyukjinKwon** [comment]
> do we want to do the doctests now instead?

I think showing a working example in doctest is the right way to use doctest but I don't mind adding the tests into separate regular tests as we did before since we're already doing that.

**Comment by HyukjinKwon** [comment]
Closing this due to author's inactivity.

**Pull request closed without merging**